### PR TITLE
Show properties for goto-diff

### PR DESCRIPTION
**Pull request by peterschrammel**
Pre-work:
- Replace cout in show_properties
- Factor out show properties command line def and docs

Show properties for goto-diff:
- Expose conversion of properties for a goto program into a JSON array so that it can be used in goto-diff
- Integrate into goto-diff to show properties for each new/modified/deleted function

Post-work:
- Enable goto check an cover options in goto-diff to have some properties to show
- Use CPROVER exit codes in goto-diff
- Add a test for goto-diff show properties

### CODE REVIEW

File: src/goto-diff/goto_diff_parse_options.cpp [hunk 1]
```diff
@@ -459,14 +484,16 @@ bool goto_diff_parse_optionst::process_goto_program(
     return true;
   }
 
-  catch(int)
+  catch(int e)
   {
+    error() << "Numeric exception : " << e << eom;
```
Comment (tautschnig) [inline src/goto-diff/goto_diff_parse_options.cpp]: Nit pick: I don't think there should be a whitespace before `:`?

File: src/goto-diff/goto_diff_parse_options.h [hunk 2]
```diff
@@ -27,6 +30,8 @@ class optionst;
 #define GOTO_DIFF_OPTIONS \
   "(json-ui)" \
   OPT_SHOW_GOTO_FUNCTIONS \
+  OPT_GOTO_CHECK \
+  "(cover):" \
```
Comment (tautschnig) [inline src/goto-diff/goto_diff_parse_options.h]: Why is it not necessary to add any of the Java stuff that got added to the help output in here?

File: src/goto-diff/goto_diff_base.cpp [hunk 3]
```diff
@@ -11,95 +11,134 @@ Author: Peter Schrammel
...
+      convert_function_group_json(
+        json_result["deletedFunctions"].make_array(),
+        deleted_functions, goto_model1);
+      result() << json_result;
       break;
     }
     case ui_message_handlert::uit::XML_UI:
     {
-      out << "not supported yet";
+      error() << "not supported yet" << eom;
```
Comment (tautschnig) [inline src/goto-diff/goto_diff_base.cpp]: Would you mind changing the error message as well to "XML output not yet supported"?

File: src/goto-diff/goto_diff_base.cpp [hunk 4]
```diff
@@ -11,95 +11,134 @@ Author: Peter Schrammel
...
+  namespacet ns(goto_model.symbol_table);
+  const symbolt &symbol = ns.lookup(function_name);
+  result() << "  " << symbol.location.get_file() << ": " << function_name << "\n";
+  if(options.get_bool_option("show-properties"))
   {
-    const symbolt &symbol =
-      namespacet(goto_model1.symbol_table).lookup(function_name);
-    result["sourceLocation"] = json(symbol.location);
+    const goto_programt &goto_program = goto_function_it->second.body;
+      for(const auto &ins : goto_program.instructions)
```
Comment (tautschnig) [inline src/goto-diff/goto_diff_base.cpp]: The `for` loop seems to be indented more than the  preceding line? Also: may I suggest to move the `const auto goto_function_it = ...` in here as it is not necessary for the other output just above?

File: src/goto-diff/goto_diff_base.cpp [hunk 5]
```diff
@@ -11,95 +11,134 @@ Author: Peter Schrammel
...
+  const auto goto_function_it =
+    goto_model.goto_functions.function_map.find(function_name);
+  PRECONDITION(goto_function_it != goto_model.goto_functions.function_map.end());
+
+  namespacet ns(goto_model.symbol_table);
   result["name"]=json_stringt(id2string(function_name));
+
+  const goto_programt &goto_program = goto_function_it->second.body;
+  const symbolt &symbol = ns.lookup(function_name);
+  result["sourceLocation"] = json(symbol.location);
```
Comment (tautschnig) [inline src/goto-diff/goto_diff_base.cpp]: The order of instructions in this function seems to be almost random.  Here's what I would do:

````c++
const namespacet ns(goto_model.symbol_table);
const symbolt &symbol = ns.lookup(function_name);

result["name"]=json_stringt(id2string(function_name));
result["sourceLocation"] = json(symbol.location);

if(options.get_bool_option("show-properties"))
{
  const auto goto_function_it = goto_model.goto_functions.function_map.find(function_name);
  ...
````

File: src/goto-diff/goto_diff_parse_options.cpp [hunk 6]
```diff
@@ -234,6 +236,12 @@ void goto_diff_parse_optionst::get_command_line_options(optionst &options)
             << " must not be given together" << eom;
     exit(1);
   }
+
+  if(cmdline.isset("show-properties"))
+    options.set_option("show-properties", true);
+  else
+    options.set_option("show-properties", false);
+
```
Comment (tautschnig) [inline src/goto-diff/goto_diff_parse_options.cpp]: `options.set_option("show-properties", cmdline.isset("show-properties"));` and no need for blank line before closing `}`

**Review by tautschnig — APPROVED** [review]

**Review by chrisr-diffblue — APPROVED** [review]
Looks ok to me.

**Pull request merged**